### PR TITLE
XTypes: MetaStruct

### PR DESCRIPTION
**Pull request by ClaytonCalabrese**
So far mutable should be done/close to done.  Putting up the PR so people can look over existing code.

### CODE REVIEW

File: dds/idl/dds_generator.h [hunk 1]
```diff
@@ -441,6 +441,73 @@ std::string wrapPrefix(AST_Type* type, WrapDirection wd)
...
+      size = 4;
+      return "ACE_CDR::ULong";
+    case AST_PredefinedType::PT_longlong:
+      size = 8;
+      return "ACE_CDR::LongLong";
+    case AST_PredefinedType::PT_ulonglong:
+      size = 8;
+      return "ACE_CDR::ULongLong";
+    case AST_PredefinedType::PT_short:
+      size = 2;
```
Comment (jwillemsen) [inline dds/idl/dds_generator.h:478]: What about using https://github.com/DOCGroup/ACE_TAO/blob/754e5a7a4d7c445d52b5376323bf19a6a206dc23/ACE/ace/CDR_Base.h#L77

File: tests/DCPS/MetaStruct/MetaStructTest.cpp [hunk 2]
```diff
@@ -69,12 +69,12 @@ bool checkVal(const T& lhs, const T& rhs, const char* name)
...
 template<typename T, typename T2>
 bool check(const T& lhs, const T& rhs, const char* name, ACE_Message_Block* amb, Value::Type type,
-          const MetaStruct& ms, T2 Value::*ptrmbr)
+          const MetaStruct& ms, T2 Value::*ptrmbr, Encoding::Kind e)
 {
   if (!checkVal(lhs, rhs, name)) return false;
 
   Message_Block_Ptr mb (amb->duplicate());
-  const Encoding encoding(Encoding::KIND_UNALIGNED_CDR);
+  const Encoding encoding(e);
```
Comment (iguessthislldo) [inline tests/DCPS/MetaStruct/MetaStructTest.cpp:77]: Woops. Missed that. Thanks for fixing it.

File: dds/idl/dds_generator.h (hunk 1 above)
Comment (iguessthislldo) [reply dds/idl/dds_generator.h:478]: Except for the types, I don't think we're using anything from ACE's CDR in our serialization system, so it's not really a time to start. I'd rather us use the non-enum constants in `dds/DCPS/Serializer.h` anyway.

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

File: dds/idl/marshal_generator.cpp [hunk 3]
```diff
@@ -3289,6 +3291,222 @@ bool marshal_generator::gen_struct(AST_Structure* node,
...
+      "    }\n";
+  }
+  //The following is Appendable/Final
+  //It is also used when Mutable but not in XCDR1 or XCDR2
+  expr = "";
+  for (Fields::Iterator i = fields.begin(); i != fields_end; ++i) {
+    AST_Field* const field = *i;
+    ACE_CDR::ULong default_id = i.pos();
+    size_t size = 0;
+    const ACE_CDR::ULong id = be_global->get_id(field, auto_id, default_id);
```
Comment (iguessthislldo) [inline dds/idl/marshal_generator.cpp:3439]:
```suggestion
    size_t size = 0;
```

**Review by iguessthislldo — APPROVED** [review]

**Pull request merged**